fix repo and added changes

Glebuska opened the pull request.

Commit:
FindMPI: Add support for the Fujitsu compiler wrappers

Co-Author: Yuichiro Utsumi <utsumi.yuichiro@jp.fujitsu.com>

Commit:
FindOpenMP: Add support for Fujitsu compilers

Co-Author: Yuichiro Utsumi <utsumi.yuichiro@jp.fujitsu.com>

Glebuska committed:
link commands added

vadim-volodin-h commented on Source/cmGlobalUnixMakefileGenerator3.h:
```diff
@@ -248,6 +251,9 @@ class cmGlobalUnixMakefileGenerator3 : public cmGlobalCommonGenerator
   size_t CountProgressMarksInAll(const cmLocalGenerator& lg);
 
   std::unique_ptr<cmGeneratedFileStream> CommandDatabase;
+  //GLEB CHANGES
```
Well, seems like you can remove this line

vadim-volodin-h reviewed (commented):
You can find a problem definition here https://gitlab.kitware.com/vad.e.volodin/cmake/-/issues/1

Please note that more tests for big repos such as opencv, lapacke, xgboost, gtest, etc. are in this repository
https://github.com/Software-Analysis-Team/link-commands
They will not be included in original cmake not to overhead it.

vadim-volodin-h commented on Source/cmMakefileLibraryTargetGenerator.cxx:
```diff
@@ -904,7 +909,8 @@ void cmMakefileLibraryTargetGenerator::WriteLibraryRules(
     const char* name = (relink ? "relink.txt" : "link.txt");
-    this->CreateLinkScript(name, real_link_commands, commands1, depends);
+      this->CreateLinkScript(name, real_link_commands, commands1, depends);
+      this->CreateLinkScriptJSON(name, files, real_link_commands);
```
What happens in case files are empty since useArchiveRules = false? There will be a command with empty files? Could you please provide an example for such case?

PolyProgrammist replied:
Well, seems like I had a mistake in November. useArchiveRules and empty files are not an issue.

PolyProgrammist replied:
But what if useLinkScript is false? Seems like cmake should build files, however the information about a command will not present in link_commands.json. Is that right? What is useLinkScript? In what cases it can be false? What happens if it is false?

vadim-volodin-h commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,7 +169,45 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+                         << "\",\n"
+                         << R"(  "files": [)";
+
+   for (unsigned long i = 0; i < sourceFiles.size(); i++){
```
use size_t type

vadim-volodin-h commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,7 +169,45 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+
+   for (unsigned long i = 0; i < sourceFiles.size(); i++){
+       *this->LinkCommandDatabase << "\"" << cmGlobalGenerator::EscapeJSON(sourceFiles[i]) << "\"";
+       if (i < sourceFiles.size() - 1) { *this->LinkCommandDatabase << ", ";
```
Please, move body code to a separate line

vadim-volodin-h commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,7 +169,45 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+                         << R"(  "files": [)";
+
+   for (unsigned long i = 0; i < sourceFiles.size(); i++){
+       *this->LinkCommandDatabase << "\"" << cmGlobalGenerator::EscapeJSON(sourceFiles[i]) << "\"";
```
Please, use separate line for EscapeJson

vadim-volodin-h commented on MY_TESTS/main.py:
```diff
@@ -0,0 +1,81 @@
+
+    for elem in link_data:
+        os.chdir(cur_dir)
+        build = re.findall(r'build\S*', elem.get('directory'))[0]
```
Use just elem.get('directory') instead of using regexp if possible

vadim-volodin-h replied:
Then just do os.makedirs(exist_ok=True) instead of iterations

vadim-volodin-h commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1800,14 +1800,59 @@ void cmMakefileTargetGenerator::CloseFileStreams()
+void cmMakefileTargetGenerator::CreateLinkScriptJSON(const char *name, const std::vector<std::string>& files,
+                                                     const std::vector<std::string> &link_commands) {
+
+    // Create the link script file.
```
This line is not creating a file

vadim-volodin-h commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1800,14 +1800,59 @@ void cmMakefileTargetGenerator::CloseFileStreams()
+    std::string delimiter = " ";
+    size_t pos = 0;
+    std::string token;
+    for (auto file : files){
```
const &

Glebuska replied:
> const &

it's impossible in this case. I change value using erase method

Glebuska replied:
> const &

I cant use it. File is mutable variable

vadim-volodin-h commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1800,14 +1800,59 @@ void cmMakefileTargetGenerator::CloseFileStreams()
+            resultStringArr.push_back(cmStrCat(workingDirectory, '/', token));
+            file.erase(0, pos + delimiter.size());
+        }
+        resultStringArr.push_back(cmStrCat(workingDirectory, '/', file));;
```
use emplace_back

vadim-volodin-h commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1800,14 +1800,59 @@ void cmMakefileTargetGenerator::CloseFileStreams()
   std::string linkScriptName =
     cmStrCat(this->TargetBuildDirectoryFull, '/', name);
+
+  std::string workingDirectory = cmSystemTools::CollapseFullPath(
```
Do you use this variable somehow after definition?

vadim-volodin-h commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -497,8 +498,10 @@ void cmMakefileExecutableTargetGenerator::WriteExecutableRule(bool relink)
+  std::string buildObjs;
   // Expand the rule variables.
   {
+      cmRulePlaceholderExpander::RuleVariables vars;
```
How do you win moving this line to the top?

Commit:
Fujitsu: Add support for the Fujitsu compiler in Trad mode

Co-Author: Yuichiro Utsumi <utsumi.yuichiro@jp.fujitsu.com>

Commit:
Tests: Update for the Fujitsu compiler

Commit:
FujitsuClang: Add support for the Fujitsu compiler in Clang mode

This should be front end compatible with vanilla clang but giving it a
unique identifier allows a project to pass additional options unique to
Fujitsu and outside the scope of a CMake builtin.

Commit:
Help: Add release notes for Fujitsu compiler support

Commit:
FindBLAS/LAPACK: Add support for the Fujitsu SSL2 library

This also does some additional work to fix issues with
libraries provided only via compiler options and no explicit
library names.

Co-Author: Yuichiro Utsumi <utsumi.yuichiro@jp.fujitsu.com>

MarkaPola committed:
UseSWIG: Run using policy settings from includer

Backport commit de7f0aa6c0 (UseSWIG: avoid spurious policy warnings,
2021-02-11) to the 3.20 release branch.  Projects need to be able to
control policies.

Issue: #22029

bradking committed:
Tests: Add RunCMake helper to run a plain script

Many tests covering CMake scripting commands can work without
initializing a full generator.

Commit:
Tests: Update for the FujitsuClang compiler

PolyProgrammist commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,6 +169,46 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+                         << R"(  "command": ")"
+                         << cmGlobalGenerator::EscapeJSON(linkCommand)
+                         << "\",\n"
+                         << R"(  "files": [)";
```
Place newline here, please
```
 "files": ["/root/lapack/build/BLAS/SRC/BLAS/SRC/CMakeFiles/blas.dir/isamax.f.o",
```
This will look better:
```
 "files": [
        "/root/lapack/build/BLAS/SRC/BLAS/SRC/CMakeFiles/blas.dir/isamax.f.o",
```

PolyProgrammist commented on .gitignore:
```diff
@@ -12,3 +12,7 @@ CMakeUserPresets.json
 # Visual Studio Code
 .vscode/
+
+#Idea Code
```
Please, remove any changes from .gitignore. Use .git/info/exclude

PolyProgrammist commented on Source/cmGlobalGenerator.cxx:
```diff
@@ -3233,7 +3233,10 @@ std::string cmGlobalGenerator::EscapeJSON(const std::string& s)
     }
   }
-  return result;
+    if (result[result.size() - 1] == ' ') {
```
Use string::back()

PolyProgrammist commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -259,7 +264,10 @@ void cmMakefileExecutableTargetGenerator::WriteNvidiaDeviceExecutableRule(
     // Use a link script.
     const char* name = (relink ? "drelink.txt" : "dlink.txt");
-    this->CreateLinkScript(name, real_link_commands, commands1, depends);
+      this->CreateLinkScript(name, real_link_commands, commands1, depends);
```
Aaaa, tabs. Please look at all the changed code and fix tabs and other formatting issues

PolyProgrammist commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -136,7 +136,12 @@ void cmMakefileExecutableTargetGenerator::WriteNvidiaDeviceExecutableRule(
   // Build list of dependencies.
   std::vector<std::string> depends;
+
+  // Build other list of dependencies, which will use to get json's links command
```
Please, include a small description of what is the "other" list, how is it different to previous

PolyProgrammist replied:
which will **be used** to get json's link command**s**

PolyProgrammist commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1748,11 +1748,50 @@ void cmMakefileTargetGenerator::AppendObjectDepends(
+        std::vector<std::string>& depends, const std::string& linkLanguage)
+{
+    // Add dependencies on the compiled object files.
+    std::string const& fullPath =
```
Specify the meaning of what is the path in the name. Is it a path to project build directory or something else? In this case, name it buildDirFullPath

PolyProgrammist commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1748,11 +1748,50 @@ void cmMakefileTargetGenerator::AppendObjectDepends(
+            this->LocalGenerator->ConvertToFullPath(
+                    this->LocalGenerator->GetHomeRelativeOutputPath());
+    for (std::string const& obj : this->Objects) {
+        std::string objTarget = cmStrCat(fullPath, obj);
```
Is this always the right behaviour? I think fullPath can be changed according to the working directory from where the object is built. I am not sure it is possible in cmake, however this may be a source of issue with lapack

Glebuska replied:
fixed

PolyProgrammist commented on Tests/LinkCommandsTest/README.md:
````diff
@@ -0,0 +1,8 @@
+```
+python3 main.py --i="path-to-project"
````
Could you please move it out to this repository https://github.com/Software-Analysis-Team/link-commands

PolyProgrammist commented on Tests/LinkCommandsTest/delete_script.txt:
```diff
@@ -0,0 +1,5 @@
+find . -type f -not \( -name '*o' -or -name '*json' -or -name 'ade' -or -name ippicv \) -delete
```
Could you please move it out to this repository https://github.com/Software-Analysis-Team/link-commands

PolyProgrammist commented on Tests/LinkCommandsTest/main.py:
```diff
@@ -0,0 +1,219 @@
+import argparse
```
Could you please move it out to this repository https://github.com/Software-Analysis-Team/link-commands

friendlyanon committed:
cmake-presets: Make generator and binaryDir fields optional

In v3 of the presets, generator and buildDir can be omitted to fall
back to regular cmake behavior when these values are not explicitly
provided by the user.

Fixes: #21987

bradking committed:
GNUInstallDirs: Clarify that CMAKE_INSTALL_<dir> may be absolute

Fixes: #21991

KyleFromKitware committed:
Tests: Add test for Ninja automoc dependency cyle

bradking committed:
Ninja Multi-Config: Fix crash on custom command config with no output

With generator expressions in a custom command's `OUTPUT` and
`BYPRODUCTS`, it is possible to have no outputs at all for a particular
configuration.  Generate no rule in this case.

Fixes: #21989

do-m-en committed:
CPackRPM: correctly handle empty dir in single package mode

Enabling CPACK_RPM_DEBUGINFO_SINGLE_PACKAGE while
having an empty directory in non main component
resulted in packaging error that the directory
does not exist.

Fixes: #21951

do-m-en committed:
CPackRPM: handle scripts in debuginfo single package mode

Enabling CPACK_RPM_DEBUGINFO_SINGLE_PACKAGE while
using rpm install or erase scripts resulted in
resulted missing scripts for non default components.

Fixes: #21974

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
Merge topic 'correct_arg_parsing' into release-3.20

79d7bcaf40 cmCommandLineArgument: correctly compute next parse index

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5936

bradking committed:
Merge topic 'correct_arg_parsing'

79d7bcaf40 cmCommandLineArgument: correctly compute next parse index

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5936

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindIntl-builtin' into release-3.20

e34d64c04d FindIntl: Fix detection of intl built in to C library
cb0d1a49e6 FindIntl: Improve documentation formatting

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Jeremy Tan <jtanx@outlook.com>
Acked-by: David Adam <zanchey@ucc.gu.uwa.edu.au>
Merge-request: !5948

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindIntl-builtin'

0850160254 Merge branch 'backport-FindIntl-builtin' into FindIntl-builtin
8e342a6637 FindIntl: Fix detection of intl built in to C library
f56c1571d2 Merge branch 'backport-FindIntl-builtin' into FindIntl-builtin
e34d64c04d FindIntl: Fix detection of intl built in to C library
cb0d1a49e6 FindIntl: Improve documentation formatting

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Jeremy Tan <jtanx@outlook.com>
Acked-by: David Adam <zanchey@ucc.gu.uwa.edu.au>
Merge-request: !5948

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'commands-file-permissions'

769ff05483 Help: Clarify permission-related command options
900184616a Cleanup: Fix misspelt name of local C++ variable
635431a0c9 Tests: Check host platform instead of target for running stat
4ceb0ca59e Tests: Remove redundant files for configure_file() tests

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5945

bradking committed:
Merge topic 'commands-file-permissions' into release-3.20

769ff05483 Help: Clarify permission-related command options
900184616a Cleanup: Fix misspelt name of local C++ variable
635431a0c9 Tests: Check host platform instead of target for running stat
4ceb0ca59e Tests: Remove redundant files for configure_file() tests

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5945

bradking committed:
Merge topic 'FindDoxygen-clarify-cmake-variable-naming'

739446a9a1 FindDoxygen: clarify the CMake variable naming scheme

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Mateusz Mania <kajtkowo@gmail.com>
Merge-request: !5943

bradking committed:
Merge topic 'SWIG'

1d8e3a1e77 UseSWIG: Use standard library name conventions for csharp language

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5949

bradking committed:
Tests: Teach RunCMake to ignore incidental 'Recompacting log' ninja output

Ninja versions prior to 1.6.0 occasionally print this message.

KyleFromKitware committed:
autogen: Don't include SKIP_AUTOMOC files in depfile

SKIP_AUTOMOC files should not be included in moc processing, which
means they shouldn't be included in the depfile either. Remove them.

Fixes: #21977

Commit:
Cray: Detect Fortran compiler version patch level if available

Fixes: #21969

hhbti committed:
Android: Fix search for binutils

Set `CMAKE_SYSTEM_PROGRAM_PATH` in `Platform/Android-Initialize` instead
of `Platform/Android` so it can be used in `CMakeFindBinUtils`.

Also add the names `llvm-strip` and `llvm-ranlib` for the corresponding
tools.

kwrobot committed:
CMake Nightly Date Stamp

MarkaPola committed:
Makefiles dependencies: normalize windows paths

Fixes: #21997

bradking committed:
Utilities/Release: Add script to generate a table of files

Fixes: #22002

kwrobot committed:
KWSys 2021-03-30 (db93a594)

Code extracted from:

    https://gitlab.kitware.com/utils/kwsys.git

at commit db93a5948e331c8920ac701e35b2f7e283646cae (master).

Upstream Shortlog
-----------------

Lemures Lemniscati (2):
      9a6d6c54 Terminal: Avoid using Windows functions on Cygwin
      4c61b968 SystemTools::FileIsFullPath: Do not accept Windows paths on Cygwin

bradking committed:
Merge topic 'nmc-cc-no-output' into release-3.20

6dd89529e8 Ninja Multi-Config: Fix crash on custom command config with no output
e21a80e97d Tests: Teach RunCMake to ignore incidental 'Recompacting log' ninja output

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Kyle Edwards <kyle.edwards@kitware.com>
Merge-request: !5950

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'nmc-cc-no-output'

6dd89529e8 Ninja Multi-Config: Fix crash on custom command config with no output
e21a80e97d Tests: Teach RunCMake to ignore incidental 'Recompacting log' ninja output

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Kyle Edwards <kyle.edwards@kitware.com>
Merge-request: !5950

bradking committed:
Merge topic 'GNUInstallDirs-doc' into release-3.20

06171fd7aa GNUInstallDirs: Clarify that CMAKE_INSTALL_<dir> may be absolute

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5947

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'GNUInstallDirs-doc'

06171fd7aa GNUInstallDirs: Clarify that CMAKE_INSTALL_<dir> may be absolute

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5947

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'ninja-automoc-cycle' into release-3.20

ca7c76269b Tests: Add test for Ninja automoc dependency cyle
54ad3e4958 autogen: Don't include SKIP_AUTOMOC files in depfile

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Jörg Bornemann <joerg.bornemann@qt.io>
Acked-by: Craig Scott <craig.scott@crascit.com>
Merge-request: !5956

bradking committed:
Merge topic 'ninja-automoc-cycle'

ca7c76269b Tests: Add test for Ninja automoc dependency cyle
54ad3e4958 autogen: Don't include SKIP_AUTOMOC files in depfile

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Jörg Bornemann <joerg.bornemann@qt.io>
Acked-by: Craig Scott <craig.scott@crascit.com>
Merge-request: !5956

bradking committed:
Merge topic 'multiple-L-labels'

44ad3f0b7f ctest: Support multiple -L and -LE options to mean "AND"

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Lars Bilke <larsbilke83@googlemail.com>
Merge-request: !5329

bradking committed:
Merge topic 'ifw-depends-name-version-parsing'

a43783a08d CPackIFW: Fix parsing of name and version in component DEPENDS
f536e6f3fb CPackIFW: Remove redundant variable assignment

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5929

bradking committed:
Merge branch 'upstream-KWSys' into update-kwsys

# By KWSys Upstream
* upstream-KWSys:
  KWSys 2021-03-30 (db93a594)

bradking committed:
ci: add jq and DevIL to Debian and Fedora base images

bradking committed:
ci: Enable jq-based tests on Linux builds

bradking committed:
ci: Enable FindDevIL tests on Linux builds

kwrobot committed:
CMake Nightly Date Stamp

MarkaPola committed:
CUDA Device link: Ensure all link options are correctly formatted

Fixes: #21994

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'android-binutils' into release-3.20

61e6fc26bc Android: Fix search for binutils

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5958

bradking committed:
Merge topic 'android-binutils'

61e6fc26bc Android: Fix search for binutils

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5958

bradking committed:
Merge topic 'update-kwsys'

25ccb2c7c3 Merge branch 'upstream-KWSys' into update-kwsys
d0b9ffb630 KWSys 2021-03-30 (db93a594)

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5960

bradking committed:
Merge topic 'cpackrpm-single-package-handling-empty-dir'

d0adb2386a CPackRPM: correctly handle empty dir in single package mode

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5951

bradking committed:
Merge topic 'cray-version-patch'

258faaeef6 Cray: Detect Fortran compiler version patch level if available

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5957

bradking committed:
Merge topic 'cray-version-patch' into release-3.20

258faaeef6 Cray: Detect Fortran compiler version patch level if available

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5957

bradking committed:
Merge branch 'release-3.20'

KyleFromKitware committed:
CMakePresets.json: Add matches condition

MarkaPola committed:
FindMPI: avoid host link options to be propagated to device link step

Fixes: #21887

bradking committed:
Utilities/Release: Update file table for 3.19.2

bradking committed:
Utilities/Release: Update file table for 3.19.3

MarkaPola committed:
Genex: $<HOST_LINK:> and $<DEVICE_LINK:> must be usable in try_compile

Fixes: #22007

bradking committed:
Utilities/Release: Update file table for 3.20.0

kwrobot committed:
CMake Nightly Date Stamp

sgn committed:
Tests/ConfigSources: fix for non main stream CMAKE_BUILD_TYPE

- None is a valid CMAKE_BUILD_TYPE
- Most of distros uses None as CMAKE_BUILD_TYPE
- When CMAKE_BUILD_TYPE=None, main_other.cpp will be compiled and linked
  into final executable, this program requires some symbols only exist
  when CUSTOM_CFG_OTHER is defined.
- And CMake also allows other CMAKE_BUILD_TYPE, too, CMake documentation
  specificaly mentions funny CMAKE_BUILD_TYPE like ReLeAsE [1]

Let's define them when non main stream like None is specified as CMAKE_BUILD_TYPE.

[1]: https://cmake.org/cmake/help/v3.20/variable/CMAKE_BUILD_TYPE.html

bradking committed:
Help: CMAKE_APPLE_SILICON_PROCESSOR cannot be set in a toolchain file

`CMakeDetermineSystem` determines the host system information before
loading the toolchain file.

Issue: #22012

bradking committed:
Merge branch 'doc-CMAKE_APPLE_SILICON_PROCESSOR' into release-3.19

Merge-request: !5965

bradking committed:
Merge topic 'make-fix-deps-paths' into release-3.20

1b346350af Makefiles dependencies: normalize windows paths

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5955

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'make-fix-deps-paths'

1b346350af Makefiles dependencies: normalize windows paths

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5955

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'CUDA-device-link'

bb96984ec9 CUDA Device link: Ensure all link options are correctly formatted

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5962

bradking committed:
Merge topic 'CUDA-device-link' into release-3.20

bb96984ec9 CUDA Device link: Ensure all link options are correctly formatted

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5962

bradking committed:
Merge topic 'genex-HOST_LINK-in-try_compile' into release-3.20

6cf81efe7d Genex: $<HOST_LINK:> and $<DEVICE_LINK:> must be usable in try_compile

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5964

bradking committed:
Merge topic 'genex-HOST_LINK-in-try_compile'

6cf81efe7d Genex: $<HOST_LINK:> and $<DEVICE_LINK:> must be usable in try_compile

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5964

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'doc-CMAKE_APPLE_SILICON_PROCESSOR' into release-3.20

3f04f69733 Help: CMAKE_APPLE_SILICON_PROCESSOR cannot be set in a toolchain file

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5965

bradking committed:
Merge topic 'doc-CMAKE_APPLE_SILICON_PROCESSOR'

3f04f69733 Help: CMAKE_APPLE_SILICON_PROCESSOR cannot be set in a toolchain file

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5965

bradking committed:
Merge branch 'release-3.19'

bradking committed:
Merge branch 'release-3.19' into release-3.20

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'cpackrpm-debuginfo-single-package-install-erase-scripts'

31c184ad69 CPackRPM: handle scripts in debuginfo single package mode

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5953

bradking committed:
Merge topic 'cmake-presets-condition-regex'

5ac8b923f5 CMakePresets.json: Add matches condition

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5963

bradking committed:
Merge topic 'fujitsu-compiler-4.0-support'

8ef55dec29 Help: Add release notes for Fujitsu compiler support
4c74c86f40 FindBLAS/LAPACK: Add support for the Fujitsu SSL2 library
376c300b25 FindOpenMP: Add support for Fujitsu compilers
9e0a1cf03e FindMPI: Add support for the Fujitsu compiler wrappers
a237450948 Tests: Update for the FujitsuClang compiler
27579e9cf1 FujitsuClang: Add support for the Fujitsu compiler in Clang mode
a55feff69c Tests: Update for the Fujitsu compiler
3c867cff4a Fujitsu: Add support for the Fujitsu compiler in Trad mode

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Axel Huebl <axel.huebl@plasma.ninja>
Acked-by: Gilles Gouaillardet <gilles@rist.or.jp>
Merge-request: !5954

bradking committed:
Utilities/Release: Add deprecation fields to File Table v1

Provide a field for clients to check now so that we can deprecate
things later.

mcclure committed:
Help: Clarify IMPORTED_LOCATION documentation

Prefer the macOS term "application bundle" over just "bundle".
Put both macOS cases together.

do-m-en committed:
CPackRPM: add scriplets tags only if scripts exist

Scriplet tags should not be added to generated
spec files if scripts weren't provided as those
tags are otherwise present in generated rpm file
even if the script wasn't provided thereby
generating unneeded dependency on shell.

Fixes: #21345

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
gitlab-ci: Tell CDash when a test-ext job is done

These jobs are self-contained and so can be re-run on their own to
produce a whole new submission to CDash.  Tell CDash when each
submission is done so re-running a job replaces results rather than
accumulating them.

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindMPI-host_link'

666f689e48 FindMPI: avoid host link options to be propagated to device link step

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Robert Maynard <robertjmaynard@gmail.com>
Acked-by: Alex <leha-bot@yandex.ru>
Merge-request: !5966

bradking committed:
Merge topic 'FindMPI-host_link' into release-3.20

666f689e48 FindMPI: avoid host link options to be propagated to device link step

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Robert Maynard <robertjmaynard@gmail.com>
Acked-by: Alex <leha-bot@yandex.ru>
Merge-request: !5966

bradking committed:
Merge topic 'doc-IMPORTED_LOCATION' into release-3.20

08a25e4902 Help: Clarify IMPORTED_LOCATION documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5968

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'doc-IMPORTED_LOCATION'

08a25e4902 Help: Clarify IMPORTED_LOCATION documentation

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5968

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'ci-test-ext-done' into release-3.20

7829136a6f gitlab-ci: Tell CDash when a test-ext job is done

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !5971

bradking committed:
Merge topic 'ci-test-ext-done'

7829136a6f gitlab-ci: Tell CDash when a test-ext job is done

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !5971

bradking committed:
Merge branch 'ci-test-ext-done' into release-3.19

Merge-request: !5971

bradking committed:
Merge branch 'release-3.19'

bradking committed:
Merge branch 'release-3.19' into release-3.20

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Help: Document in add_library how to import libraries with SONAME

When using `add_library` to create an `IMPORTED` library target, there
are a few target properties that are essential to correctly tell CMake
about the library file.  We already cover `IMPORTED_LOCATION` and
`IMPORTED_IMPLIB`.  Add `IMPORTED_SONAME` and `IMPORTED_NO_SONAME`,
which are important in certain cases.

Fixes: #22016

kwrobot committed:
CMake Nightly Date Stamp

D-Barber committed:
Autogen: Rename the variable for UI files with UIC options

Monitoring for UI file changes in the target sources will require keeping track of the files without options as well so this will improve clarity.

D-Barber committed:
Autogen: Add detection of UI file changes to the pre-build VS case

This is achieved by adding a rule file which carries the UI files as dependencies but performs no meaningful command. Its output path points to a timestamp file which is instead touched by the pre-build command that runs autogen. The rule file therefore triggers the build if any of the files have been changed more recently than the last autogen run.

Fixes: #17959
Fixes: #18741

D-Barber committed:
Autogen: Move duplicated executable dependency code into a function

kwrobot committed:
CMake Nightly Date Stamp

D-Barber committed:
Help: Fix a typo in the AUTOGEN_SOURCE_GROUP help

D-Barber committed:
Help: Document the AUTOUIC_SOURCE_GROUP property

This property came into existence with the changes in b4a103bdecc437d37580d0ab540f6014006d661c because generated files are now being added for UIC.

D-Barber committed:
Autogen: Fix a bug in setting the UIC source group

If the relative path to a UI file was empty then the source group for the corresponding autogen header file would not apply because of a double slash in the path.

mathstuf committed:
FindHDF5: find include directories after the library

Some include paths might depend on the library location (primarily for
Fortran).

mathstuf committed:
FindHDF5: track the "main" library for each language

This will be used to build up additional include paths to search.

mathstuf committed:
FindHDF5: add compiler-specific module search paths

mathstuf committed:
FindHDF5: search for the new Fortran HL library name

HDF5 1.10.6 renamed this library to match the other language binding
library names.

Fixes: #20205

DerDakon committed:
Tests: bump cmake_minimum_required version to 2.8.12

This needlessly produces warnings during the test runs that no-one
sees but that are distracting when actually inspecting the logs.

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
libarchive: Use uint8_t instead of u_char

The latter is not a standard type.

Fixes: #22013

Neumann-A committed:
Help: Add Q_NAMESPACE_EXPORT to CMAKE_AUTOMOC_MACRO_NAMES default values

This was accidentally left out of commit 426941c433 (Autogen: Recognize
the new Q_NAMESPACE_EXPORT macro in AUTOMOC, 2020-02-26,
v3.17.0-rc2~3^2).

bradking committed:
Merge topic 'doc-imported-soname' into release-3.20

4d6e899578 Help: Document in add_library how to import libraries with SONAME

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5975

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'doc-imported-soname'

4d6e899578 Help: Document in add_library how to import libraries with SONAME

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5975

bradking committed:
Merge topic 'cpack-rpm-add-scriplets-tags-only-if-needed'

34c8a23044 CPackRPM: add scriplets tags only if scripts exist

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5970

bradking committed:
Merge topic 'tests-CMAKE_BUILD_TYPE-None'

ce1cadd35a Tests/ConfigSources: fix for non main stream CMAKE_BUILD_TYPE

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5967

bradking committed:
Merge branch 'backport-3.18-rel-file-table' into backport-3.19-rel-file-table

bradking committed:
Merge tag 'v3.19.2' into backport-3.19-rel-file-table

bradking committed:
Merge tag 'v3.19.3' into backport-3.19-rel-file-table

bradking committed:
Merge branch 'release-3.19' into backport-3.19-rel-file-table

bradking committed:
Merge branch 'backport-3.19-rel-file-table' into rel-file-table

bradking committed:
UseSWIG: Transform swig depfile to match Ninja generator paths

Since commit 89b01b04fa (UseSWIG: use swig tool to generate
dependencies, 2021-01-12, v3.20.0-rc1~120^2) we use a tool-provided
depfile to extract dependencies under the Ninja generator.  Enable
`CMP0116` to ensure depfile paths are translated to match what the Ninja
generator writes to the build manfiest.

Fixes: #22029

bradking committed:
Merge branch 'backport-UseSWIG-policies' into UseSWIG-policies

robertmaynard committed:
CMakeDetermineCompilerABI: Revert "Parse library arch from versioned paths"

The change in commit 657fc3a9a7 (CMakeDetermineCompilerABI: Parse
library arch from versioned paths, 2021-02-03, v3.20.0-rc1~40^2) caused
`CMAKE_LIBRARY_ARCHITECTURE` to be populated on non-multiarch platforms
if their compilers happen to use `$arch/$version` library directories.
Revert the use of versioned library paths.

Fixes: #22024

robertmaynard committed:
CMakeDetermineCompilerABI: Extract lib arch from implicit object file paths

The NVHPC compiler does not have any implicit link directories that can
be used to detect `CMAKE_LIBRARY_ARCHITECTURE`, but it does have
implicit object files.  Extract implicit object file paths from link
lines and check them for the `CMAKE_LIBRARY_ARCHITECTURE` pattern.

Issue: #22024

craigscott-crascit committed:
Apple: Set CMAKE_SHARED_LIBRARY_RUNTIME_C_FLAG on non-macOS too

Since CMake 3.19, we no longer support macOS SDKs older than 10.5,
which corresponds to Xcode 3. Supporting older Xcode versions for
device platforms is also not realistic. We therefore expect the -rpath
linker option should always be supported now.

When targeting iOS, tvOS or watchOS, the previous disabling of -rpath
support meant that the install_name_dir of shared libraries and
frameworks was unable to use @rpath. This resulted in embedding
absolute paths for their install_name. When they were embedded in an
app bundle, this would cause the app to fail at runtime. By enabling the
-rpath linker option, the default install_name_dir is now @rpath for these platforms, which results in binaries that do work at runtime.

Fixes: #20036

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
BinUtils: Avoid clobbering a variable named without a private prefix

Refactoring in commit b12aec6c8d (BinUtils: prefer bin utils matching
the compiler version extension, 2020-12-11, v3.20.0-rc1~171^2)
accidentally leaked a `TOOL` variable.  Name temporary variables with a
private prefix.

Fixes: #22035

bradking committed:
Merge topic 'add_q_namespace_export_docs' into release-3.20

3538f1c69b Help: Add Q_NAMESPACE_EXPORT to CMAKE_AUTOMOC_MACRO_NAMES default values

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5982

bradking committed:
Merge topic 'add_q_namespace_export_docs'

3538f1c69b Help: Add Q_NAMESPACE_EXPORT to CMAKE_AUTOMOC_MACRO_NAMES default values

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5982

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'libarchive-u_char'

a9fc751e65 libarchive: Use uint8_t instead of u_char

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5981

bradking committed:
Merge topic 'libarchive-u_char' into release-3.20

a9fc751e65 libarchive: Use uint8_t instead of u_char

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5981

bradking committed:
Merge branch 'libarchive-u_char' into release-3.19

Merge-request: !5981

bradking committed:
Merge branch 'release-3.19'

bradking committed:
Merge branch 'release-3.19' into release-3.20

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindHDF5-new-fortran-library-name'

ff84a80b46 FindHDF5: search for the new Fortran HL library name

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5976

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindHDF5-new-fortran-library-name' into release-3.20

ff84a80b46 FindHDF5: search for the new Fortran HL library name

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5976

bradking committed:
Merge topic 'UseSWIG-policies' into release-3.20

c3d0b25514 UseSWIG: Transform swig depfile to match Ninja generator paths
8c525d7e16 UseSWIG: Run using policy settings from includer

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5983

bradking committed:
Merge topic 'UseSWIG-policies'

05881c1cbe Merge branch 'backport-UseSWIG-policies' into UseSWIG-policies
c3d0b25514 UseSWIG: Transform swig depfile to match Ninja generator paths
8c525d7e16 UseSWIG: Run using policy settings from includer

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5983

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'nvhpc-lib-arch' into release-3.20

764606e256 CMakeDetermineCompilerABI: Extract lib arch from implicit object file paths
5d44d73bbe CMakeDetermineCompilerABI: Revert "Parse library arch from versioned paths"

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5984

bradking committed:
Merge topic 'nvhpc-lib-arch'

764606e256 CMakeDetermineCompilerABI: Extract lib arch from implicit object file paths
5d44d73bbe CMakeDetermineCompilerABI: Revert "Parse library arch from versioned paths"

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5984

bradking committed:
Merge topic 'Test_min_version'

f6b4db365a Tests: bump cmake_minimum_required version to 2.8.12

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5979

bradking committed:
Merge branch 'backport-3.18-rel-file-table' into release-3.18

Merge-request: !5961

bradking committed:
Merge branch 'release-3.18' into release-3.19

bradking committed:
Merge branch 'backport-3.19-rel-file-table' into release-3.19

bradking committed:
Merge topic 'rel-file-table'

e17006c497 Utilities/Release: Add deprecation fields to File Table v1
89ab613248 Utilities/Release: Update file table for 3.20.0
c5777e5421 Merge branch 'backport-3.19-rel-file-table' into rel-file-table
f9f59fb0b2 Merge branch 'release-3.19' into backport-3.19-rel-file-table
30959aec8f Utilities/Release: Update file table for 3.19.3
d206c8f1ec Merge tag 'v3.19.3' into backport-3.19-rel-file-table
004dbbaddd Utilities/Release: Update file table for 3.19.2
1e3301a28c Merge tag 'v3.19.2' into backport-3.19-rel-file-table
...

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Reviewed-by: Craig Scott <craig.scott@crascit.com>
Merge-request: !5961

bradking committed:
Merge topic 'rel-file-table' into release-3.20

e17006c497 Utilities/Release: Add deprecation fields to File Table v1
89ab613248 Utilities/Release: Update file table for 3.20.0
c5777e5421 Merge branch 'backport-3.19-rel-file-table' into rel-file-table
f9f59fb0b2 Merge branch 'release-3.19' into backport-3.19-rel-file-table
30959aec8f Utilities/Release: Update file table for 3.19.3
d206c8f1ec Merge tag 'v3.19.3' into backport-3.19-rel-file-table
004dbbaddd Utilities/Release: Update file table for 3.19.2
1e3301a28c Merge tag 'v3.19.2' into backport-3.19-rel-file-table
...

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Reviewed-by: Craig Scott <craig.scott@crascit.com>
Merge-request: !5961

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge branch 'release-3.19' into release-3.20

bradking committed:
Merge branch 'release-3.18'

bradking committed:
Merge branch 'release-3.19'

bradking committed:
Merge branch 'release-3.20'

robertmaynard committed:
FindThreads: Reduce amount of false positive error logging

When a CMake based project fails, users start looking at
CMakeError.log and think that the FindThread output is
the problem, but it is a false positive.

Now FindThreads only logs to `CMakeError.log` when it fails
to find a suitable flag, reducing the amount of false positives in `CMakeError.log`.

Fixes #16540, Fixes #19344

mathstuf committed:
ci: pay attention to the machine load when running tests

CMake's test suite is sensitive to machine load and can cause spurious
timeouts. To acknowledge this, prevent CTest from spawning tests into a
busy machine.

mathstuf committed:
ci: limit builds by machine load

To avoid starving test jobs, the build jobs should also back off if
there is a lot of machine load.

bradking committed:
Merge branch 'backport-ci-linux-updates' into ci-linux-updates

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'ci-linux-updates'

1035848e9c ci: Enable FindDevIL tests on Linux builds
be023ff4db Merge branch 'backport-ci-linux-updates' into ci-linux-updates
c6a45968d6 ci: Enable jq-based tests on Linux builds
8ca8b74a96 ci: add jq and DevIL to Debian and Fedora base images

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !5987

bradking committed:
Merge topic 'ci-linux-updates' into release-3.20

c6a45968d6 ci: Enable jq-based tests on Linux builds
8ca8b74a96 ci: add jq and DevIL to Debian and Fedora base images

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !5987

bradking committed:
Tests: Add RunCTest helper option to control CTest verbose flag

bradking committed:
ctest: Display test "WAITING" messages only in extra-verbose mode

These messages are incidental scheduling output.  Avoid cluttering
the normal verbose output showing tests starting and finishing.

bradking committed:
CMake 3.19.8

bradking committed:
Merge branch 'release-3.19'

bradking committed:
Merge branch 'release-3.19' into release-3.20

bradking committed:
Merge branch 'release-3.20'

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
Merge topic 'autouic-source-group-fixes'

e8ddc1e268 Autogen: Fix a bug in setting the UIC source group
7f0151dc91 Help: Document the AUTOUIC_SOURCE_GROUP property
f6c47dd0b4 Help: Fix a typo in the AUTOGEN_SOURCE_GROUP help

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Alexey Edelev <alexey.edelev@qt.io>
Merge-request: !5978

bradking committed:
Merge topic 'FindHDF5-find-fortran-modules-automatically'

cd48433e96 FindHDF5: add compiler-specific module search paths
78287ec0af FindHDF5: track the "main" library for each language
3654068e80 FindHDF5: find include directories after the library

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5977

bradking committed:
Makefiles: Fix dependency extraction with CUDA < 10.2 and host compiler

Since commit 2c71d051fa (Makefiles Generators: use compiler for
dependencies generation, 2020-10-18, v3.20.0-rc1~392^2) we invoke `nvcc`
for CUDA < 10.2 a second time in order to generate a depfile.  When
`CMAKE_CUDA_HOST_COMPILER` is set, the second invocation is missing its
`-ccbin=` option, even after refactoring in commit 8981e3e7cc
(NVIDIA-CUDA: rely on new capabilities for deps generation, 2020-12-02,
v3.20.0-rc1~362^2).

Ideally we should move the `-ccbin=` flag into `Compiler/NVIDIA-CUDA`,
but that will add `CMAKE_CUDA_HOST_COMPILER` support on Windows in
command-line generators but not the Visual Studio generators.
For now, add the flag to the depfile command specifically.

Fixes: #22037

bradking committed:
Merge topic 'ios-rpath-linker-flag' into release-3.20

4aed96e230 Apple: Set CMAKE_SHARED_LIBRARY_RUNTIME_C_FLAG on non-macOS too

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5980

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'ios-rpath-linker-flag'

4aed96e230 Apple: Set CMAKE_SHARED_LIBRARY_RUNTIME_C_FLAG on non-macOS too

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5980

bradking committed:
Merge topic 'binutils-var-private'

4d786dfcfa BinUtils: Avoid clobbering a variable named without a private prefix

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5986

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'binutils-var-private' into release-3.20

4d786dfcfa BinUtils: Avoid clobbering a variable named without a private prefix

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5986

bradking committed:
FindBLAS: Fix detection of OpenMP as dependency of BLA_STATIC

Update the change from commit f7f3d8987a (FindBLAS: Add dependency of
OpenBLAS on OpenMP for BLA_STATIC, 2020-11-10, v3.20.0-rc1~492^2):

* If C is not enabled, find CXX OpenMP libraries instead.

* Do not use BLA_STATIC's custom CMAKE_FIND_LIBRARY_SUFFIXES for OpenMP.
  It can break projects that already call `find_package(OpenMP)` and
  expect a shared library.  Whether OpenMP is static is orthogonal to
  whether BLAS is static.

Fixes: #22039
Issue: #16221

mathstuf committed:
Help/dev/review: mention `--pretty=reference`

This was added in Git 2.25 (13 Jan 2020).

bradking committed:
Restore support for backslashes in initial language-wide flags

Refactoring in commit bdc40742bd (CMakeDetermineCompilerId: Test without
COMPILER_ID_FLAGS if REQUIRE_SUCCESS, 2021-02-27, v3.20.0-rc3~6^2) added
an extra macro layer through which flag strings are passed.  That caused
an extra level of argument re-parsing, and broke flags with backslashes.
Pass flags to the helper macro through variable names instead.

Fixes: #22041

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindBLAS-default-OpenMP' into release-3.20

9ef82d95d8 FindBLAS: Fix detection of OpenMP as dependency of BLA_STATIC

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5993

bradking committed:
Merge topic 'FindBLAS-default-OpenMP'

9ef82d95d8 FindBLAS: Fix detection of OpenMP as dependency of BLA_STATIC

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5993

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'cuda-depfile-ccbin' into release-3.20

8e38985db7 Makefiles: Fix dependency extraction with CUDA < 10.2 and host compiler

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Raul Tambre <raul@tambre.ee>
Acked-by: Robert Maynard <robertjmaynard@gmail.com>
Merge-request: !5992

bradking committed:
Merge topic 'cuda-depfile-ccbin'

8e38985db7 Makefiles: Fix dependency extraction with CUDA < 10.2 and host compiler

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Raul Tambre <raul@tambre.ee>
Acked-by: Robert Maynard <robertjmaynard@gmail.com>
Merge-request: !5992

bradking committed:
Merge topic 'flags-with-backslash' into release-3.20

3953dfcb31 Restore support for backslashes in initial language-wide flags

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5995

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'flags-with-backslash'

3953dfcb31 Restore support for backslashes in initial language-wide flags

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5995

bradking committed:
Merge topic 'find_threads_only_log_on_failure'

d6da6784bf FindThreads: Reduce amount of false positive error logging

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5988

bradking committed:
Merge topic 'ctest-wait-verbosity'

82e40607a4 ctest: Display test "WAITING" messages only in extra-verbose mode
80da23f2a2 Tests: Add RunCTest helper option to control CTest verbose flag

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !5990

bradking committed:
Merge topic 'presets-optional-generator-and-builddir'

06e6981336 cmake-presets: Make generator and binaryDir fields optional

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5946

bradking committed:
Merge topic 'newer-git-show'

f71e7b8156 Help/dev/review: mention `--pretty=reference`

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5994

bradking committed:
CMake 3.20.1

bradking committed:
Merge branch 'release-3.20'

vadim-volodin-h commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -264,10 +265,11 @@ void cmMakefileExecutableTargetGenerator::WriteNvidiaDeviceExecutableRule(
-      if(this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")){
-        this->CreateLinkScriptJSON(name, real_link_commands, depends);
-      }
+    this->CreateLinkScript(name, real_link_commands, commands1, depends);
```
There are two places with this if. Please, consider both. This is first

Glebuska replied:
They call in different function. This is WriteNvidiaDeviceExecutableRule function

vadim-volodin-h commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -632,8 +635,9 @@ void cmMakefileExecutableTargetGenerator::WriteExecutableRule(bool relink)
     this->CreateLinkScript(name, real_link_commands, commands1, depends);
-    if(this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")){
-      this->CreateLinkScriptJSON(name, real_link_commands, dependsForJson);
+    if (this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")) {
```
There are two places with this if. Please, consider both. This is second

Glebuska replied:
This is WriteExecutableRule and other...

vadim-volodin-h commented on Source/cmMakefileLibraryTargetGenerator.cxx:
```diff
@@ -404,8 +404,9 @@ void cmMakefileLibraryTargetGenerator::WriteNvidiaDeviceLibraryRules(
     this->CreateLinkScript(name, real_link_commands, commands1, depends);
-    if(this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")){
-       this->CreateLinkScriptJSON(name, real_link_commands, dependsForJson);
+    if (this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")) {
```
Well, one more place

Glebuska replied:
one more place - one more different function :)

vadim-volodin-h commented on Source/cmMakefileLibraryTargetGenerator.cxx:
```diff
@@ -915,8 +918,9 @@ void cmMakefileLibraryTargetGenerator::WriteLibraryRules(
     this->CreateLinkScript(name, real_link_commands, commands1, depends);
-    if(this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")){
-      this->CreateLinkScriptJSON(name, real_link_commands, dependsForJson);
+    if (this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")) {
```
Wow, the fourth

PolyProgrammist commented on Source/cmGlobalGenerator.cxx:
```diff
@@ -3233,7 +3233,10 @@ std::string cmGlobalGenerator::EscapeJSON(const std::string& s)
     }
   }
-  return result;
+    if (result.back() == ' ') {
```
Please, move tabulation back

PolyProgrammist commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,6 +169,47 @@ void cmGlobalUnixMakefileGenerator3::Generate()
     this->CommandDatabase.reset();
   }
+
+    if (this->LinkCommandDatabase) {
```
Please, move tabulation back

PolyProgrammist commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1748,11 +1748,48 @@ void cmMakefileTargetGenerator::AppendObjectDepends(
+        this->GeneratorTarget->GetModuleDefinitionInfo(
+          this->GetConfigName())) {
+    for (cmSourceFile const* src : mdi->Sources) {
+      depends.push_back(src->GetFullPath());
```
emplace_back

PolyProgrammist commented on Source/cmMakefileTargetGenerator.cxx:
```diff
@@ -1748,11 +1748,48 @@ void cmMakefileTargetGenerator::AppendObjectDepends(
+  std::vector<cmSourceFile const*> manifest_srcs;
+  this->GeneratorTarget->GetManifests(manifest_srcs, this->GetConfigName());
+  for (cmSourceFile const* manifest_src : manifest_srcs) {
+    depends.push_back(manifest_src->GetFullPath());
```
emplace_back, and everywhere else

PolyProgrammist commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+  void ParseTranslationUnits()
+  {
+    this->TranslationUnits = TranslationUnitsType();
+    ExpectOrDie('[', "at start of compile command file\n");
```
link command file

PolyProgrammist commented on Tests/LinkCommandOutput/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+add_executable(maina maina.cpp)
+add_executable(mainb mainb.cpp b.cpp)
+include_directories(${LinkCommandOutput_SOURCE_DIR}/../../Source)
+target_link_libraries(maina PRIVATE say-c say-a)
```
Please add newline to end of file

PolyProgrammist replied:
![image](https://user-images.githubusercontent.com/17502570/114062818-9c55f380-98a0-11eb-8112-1bd43db5693b.png)

PolyProgrammist replied:
See that minus in a circle - it says you don't have new line. Please check them here https://github.com/Software-Analysis-Team/CMake/pull/3/files

kwrobot committed:
CMake Nightly Date Stamp

vadim-volodin-h commented on Source/cmGlobalGenerator.cxx:
```diff
@@ -3233,6 +3233,9 @@ std::string cmGlobalGenerator::EscapeJSON(const std::string& s)
         result += i;
     }
   }
+  if (result.back() == ' ') {
```
test resolve

rbkn99 commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -613,6 +630,11 @@ void cmMakefileExecutableTargetGenerator::WriteExecutableRule(bool relink)
     this->LocalGenerator->SetLinkScriptShell(false);
   }
 
+  if (this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")) {
```
Duplicated code: https://github.com/Software-Analysis-Team/CMake/blob/9dcad0c51f680b08f53f75f260934cc53a03b59c/Source/cmMakefileExecutableTargetGenerator.cxx#L263-L266 https://github.com/Software-Analysis-Team/CMake/blob/9dcad0c51f680b08f53f75f260934cc53a03b59c/Source/cmMakefileLibraryTargetGenerator.cxx#L917-L920 https://github.com/Software-Analysis-Team/CMake/blob/9dcad0c51f680b08f53f75f260934cc53a03b59c/Source/cmMakefileLibraryTargetGenerator.cxx#L402-L405
Maybe move it to a separate method?

rbkn99 commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+
+#include "cmSystemTools.h"
+
+class LinkCommandParser
```
Why you don't use existing JSON parsers? See https://github.com/nlohmann/json for example.

PolyProgrammist replied:
Yes, that's reasonable. Please take a look at json utilities https://github.com/Software-Analysis-Team/CMake/tree/master/Utilities/cmjsoncpp

PolyProgrammist replied:
I am not sure nlohmann/json could be easily approved in cmake, but please consider using an existing json parser

operasfantom commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+class LinkCommandParser
+{
+public:
+  class CommandType : public std::map<std::string, std::string>
```
As for me, public inheriting from STL class doesn't make sense in that case. Firstly, `std::map` doesn't have virtual methods and you actually hide `at()`. Secondly, you only use `at()` and `operator[]` there. 
I would suggest either to create private field with type `std::map<std::string, std::string>` or use private inheritance instead.

PolyProgrammist commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,6 +169,47 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+      "/link_commands.json";
+    this->LinkCommandDatabase =
+      cm::make_unique<cmGeneratedFileStream>(commandDatabaseName);
+    *this->LinkCommandDatabase << "[\n";
```
Also guys suggest to use json library for output

operasfantom commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+
+  bool IsWhitespace()
+  {
+    return (this->C == ' ' || this->C == '\t' || this->C == '\n' ||
```
I believe, that the similar function has already been written in utils or anywhere else. Or at the very least, you may use std::isspace, which works for different locales.

operasfantom commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+
+  void ErrorExit(const std::string& message)
+  {
+    std::cout << "ERROR: " << message;
```
Should it be printed in `std::cerr` instead?

operasfantom commented on Source/cmMakefileExecutableTargetGenerator.cxx:
```diff
@@ -254,6 +260,11 @@ void cmMakefileExecutableTargetGenerator::WriteNvidiaDeviceExecutableRule(
   }
 
+  if (this->Makefile->IsOn("CMAKE_EXPORT_LINK_COMMANDS")) {
+    this->CreateLinkScriptJSON(real_link_commands, depends,
```
Why do not you use `dependsForJson` there?

operasfantom commented on Source/cmGlobalUnixMakefileGenerator3.cxx:
```diff
@@ -169,6 +169,47 @@ void cmGlobalUnixMakefileGenerator3::Generate()
+  }
+}
+
+void cmGlobalUnixMakefileGenerator3::AddCXXLinkCommand(const std::vector<std::string>& sourceFiles, const std::string& workingDirectory,
```
Why not to just name that as "AddLinkCommand"?

operasfantom commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,164 @@
+  class CommandType : public std::map<std::string, std::string>
+  {
+  public:
+    std::string const& at(std::string const& k) const
```
The semantic of you method is different from that than one in `std::map`. Specifically, `std::map::at` throws an exceptions if key doesn't exist. Could you rename you method?

MarkaPola committed:
Genex: add_custom_command: DEPFILE supports genex

This facility is very useful for 'Ninja Multi-Config' and required
as well for future support of DEPFILE in 'Xcode' and 'Visual Studio'
generators (#20286).

PolyProgrammist commented on Source/cmMakefileLibraryTargetGenerator.cxx:
```diff
@@ -690,6 +704,7 @@ void cmMakefileLibraryTargetGenerator::WriteLibraryRules(
   }
 
   // Expand the rule variables.
+  std::set<int> ranlibIndex;
```
Please, add a comment on what is this variable and how is it used

kwrobot committed:
CMake Nightly Date Stamp

D-Barber committed:
Autogen: Move duplicated filename generation code into a function

kwrobot committed:
CMake Nightly Date Stamp

sizeofvoid committed:
OpenBSD: Fix system feature definitions

Since commit f034b0f663 (CMake compilation: do not use compiler
extensions, 2020-03-14, v3.18.0-rc1~494^2), some sources explicitly
enable needed system APIs on some platforms using definitions like
`_POSIX_C_SOURCE` and `_XOPEN_SOURCE`.  Drop the definitions for
OpenBSD, which provides the POSIX APIs by default.

kwrobot committed:
CMake Nightly Date Stamp

kwrobot committed:
CMake Nightly Date Stamp

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
Merge topic 'ci-parallelism-limits'

3fd2cffb4d ci: limit builds by machine load
6317f92120 ci: pay attention to the machine load when running tests

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5989

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'ci-parallelism-limits' into release-3.20

3fd2cffb4d ci: limit builds by machine load
6317f92120 ci: pay attention to the machine load when running tests

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5989

bradking committed:
Merge topic 'vs-autouic-on-ui-file-change'

a8d879cf45 Autogen: Add detection of UI file changes to the pre-build VS case
1876f2d03f Autogen: Move duplicated filename generation code into a function
087b717aef Autogen: Move duplicated executable dependency code into a function
0656bebeae Autogen: Rename the variable for UI files with UIC options

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5999

bradking committed:
Merge topic 'add_custom_command-DEPFILE-genex'

0c47b91fcc Genex: add_custom_command: DEPFILE supports genex

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !5998

bradking committed:
Merge topic 'openbsd-defines'

317a477283 OpenBSD: Fix system feature definitions

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !6000

bradking committed:
Merge topic 'openbsd-defines' into release-3.20

317a477283 OpenBSD: Fix system feature definitions

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !6000

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Help: Add 3.20 release note for error on unknown arguments

This was left out of commit 1b6c5333a0 (cmake: Error out on unknown
arguments starting with `-`., 2020-11-30, v3.20.0-rc1~370^2).

Fixes: #22060

kwrobot committed:
KWSys 2021-04-14 (7ab1c601)

Code extracted from:

    https://gitlab.kitware.com/utils/kwsys.git

at commit 7ab1c60177e76707fbdd462fdbf7304ad2532edc (master).

Upstream Shortlog
-----------------

Brad King (11):
      85e76a10 Status: Add class to hold an OS-specific status of a system operation
      eec0add2 Directory: Return Status from Load
      955e7c1e SystemTools: Return Status from RemoveFile and RemoveADirectory
      c5a99e59 SystemTools: Return Status from GetPermissions and SetPermissions
      fadfd7a0 SystemTools: Return Status from Touch
      06216f0b SystemTools: Return Status from FileTimeCompare
      ab198a22 SystemTools: Return Status from GetShortPath
      2446b205 SystemTools: Return Status from MakeDirectory
      6aec6af9 SystemTools: Return Status from ChangeDirectory
      0664553b SystemTools: Return Status from copy operations
      a95f61cf SystemTools: Return Status from symlink operations

bradking committed:
Merge branch 'upstream-KWSys' into update-kwsys

# By KWSys Upstream
* upstream-KWSys:
  KWSys 2021-04-14 (7ab1c601)

bradking committed:
Update CMake code using KWSys to account for Status return values

KWSys as of 2021-04-14 changed the return type of `SystemTools`
operations from `bool` to `Status`.  Update our call sites.
This may improve error reporting accuracy in a few places.

bradking committed:
cmSystemTools: Simplify using KWSys Status

bradking committed:
Find{BLAS,LAPACK}: Revert bad refactoring of internal CHECK_*_LIBRARIES

Refactoring in commit 4c74c86f40 (FindBLAS/LAPACK: Add support for the
Fujitsu SSL2 library, 2021-01-27) was done in order to support calling
`find_library` on the dependencies as well as the candidate libraries.
However, it broke a few things:

* Intel MKL's BLAS/LAPACK are no longer found.  We specify their
  dependencies using `-l...` flags, so we should not try to use
  `find_library` for them.

* The dependencies are repeated because we accumulate them in the
  `find_library` search loop and then append them at the end too.

Revert the incorrect part of the refactoring.  Retain the flags part
needed for the Fujitsu vendor.

Fixes: #22056

bradking committed:
Find{BLAS,LAPACK}: Add test case covering Intel MKL

bradking committed:
ci: factor out Intel Compiler common configure script

Share common content between the intelclassic and inteloneapi variants.

bradking committed:
Find{BLAS,LAPACK}: Simplify appending to list of libraries

bradking committed:
Find{BLAS,LAPACK}: Clarify name of internal argument for dependencies

bradking committed:
Find{BLAS,LAPACK}: Convert internal CHECK_*_LIBRARIES to functions

bradking committed:
Find{BLAS,LAPACK}: Generalize recognition of dependencies as link flags

bradking committed:
Find{BLAS,LAPACK}: Make library variable names more robust

kwrobot committed:
CMake Nightly Date Stamp

bradking committed:
Merge topic 'doc-error-unknown-arg'

06d26c7baf Help: Add 3.20 release note for error on unknown arguments

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !6003

bradking committed:
Merge topic 'doc-error-unknown-arg' into release-3.20

06d26c7baf Help: Add 3.20 release note for error on unknown arguments

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !6003

bradking committed:
Merge branch 'release-3.20'

bradking committed:
Merge topic 'FindBLAS-FindLAPACK'

dcd604ee62 Find{BLAS,LAPACK}: Make library variable names more robust
98ef6632d9 Find{BLAS,LAPACK}: Generalize recognition of dependencies as link flags
a57c4eef76 Find{BLAS,LAPACK}: Convert internal CHECK_*_LIBRARIES to functions
a5a6ac7033 Find{BLAS,LAPACK}: Clarify name of internal argument for dependencies
d248401d12 Find{BLAS,LAPACK}: Simplify appending to list of libraries
e4649d1d15 Find{BLAS,LAPACK}: Add test case covering Intel MKL
116edb5c04 Find{BLAS,LAPACK}: Revert bad refactoring of internal CHECK_*_LIBRARIES
44bcec240b ci: factor out Intel Compiler common configure script

Acked-by: Kitware Robot <kwrobot@kitware.com>
Merge-request: !6005

bradking committed:
Merge topic 'update-kwsys'

3ef5dab010 cmSystemTools: Simplify using KWSys Status
ec1b6157cb Update CMake code using KWSys to account for Status return values
eef585efaa Merge branch 'upstream-KWSys' into update-kwsys
a790167101 KWSys 2021-04-14 (7ab1c601)

Acked-by: Kitware Robot <kwrobot@kitware.com>
Acked-by: Ben Boeckel <ben.boeckel@kitware.com>
Merge-request: !6004

PolyProgrammist commented on Tests/CMakeLib/run_link_commands.cpp:
```diff
@@ -0,0 +1,38 @@
+#include <utility>
+#include <vector>
+
+#include "../Utilities/cmjsoncpp/include/json/json.h"
```
Please try to modify path so that it is easier to read. For example, something like this:
#include <json/json.h>
https://github.com/Kitware/CMake/blob/859241d2bbaae83f06c44bc21ab0dbd38725a3fc/Tests/FindJsonCpp/Test/main.cxx
